File: solutions/typescript/bank-account/1/bank-account.ts
//
// This is only a SKELETON file for the 'Bank Account' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class ValueError extends Error {
  constructor() {
    super('Bank account error')
  }
}

export class BankAccount {
  #balance: number = 0;
  #status: string = 'uninitialized';

  open(): void {
    this.#checkStatus('open');
    this.#balance = 0;
    this.#status = 'open';
  }

  close(): void {
    this.#checkStatus('close');
    this.#status = 'closed';
  }

  deposit(amt: number): void {
    this.#checkStatus();
    this.#checkAmt(amt);
    this.#balance += amt;
  }

  withdraw(amt: number): void {
    this.#checkStatus();
    this.#checkAmt(amt);
    this.#checkBalance(amt);
    this.#balance -= amt;
  }

  get balance(): number {
    this.#checkStatus();
    return this.#balance;
  }

  #checkStatus(operation: string = ''): void {
    if ((operation == '') && this.#status === 'closed'
      || (operation == 'open' && (! ['uninitialized', 'closed'].includes(this.#status)))
      || (operation == 'close' && this.#status !== 'open')) {
      throw new ValueError();
    }
  }

  #checkBalance(amt: number): void {
    if (this.#balance < amt) {
      throw new ValueError();
    }
  }

  #checkAmt(amt: number): void {
    if (amt < 0) {
      throw new ValueError();
    }
  }
}

File: solutions/ruby/bank-account/1/bank_account.rb
class BankAccount
    # Initialize instance variables
    def initialize
      @balance = 0
      @status = 'uninitialized'
    end
  
    # Open the bank account
    def open
      check_status('open')
      @balance = 0
      @status = 'open'
    end
  
    # Close the bank account
    def close
      check_status('close')
      @status = 'closed'
    end
  
    # Deposit an amount into the bank account
    def deposit(amount)
      check_status
      check_amount(amount)
      @balance += amount
    end
  
    # Withdraw an amount from the bank account
    def withdraw(amount)
      check_status
      check_amount(amount)
      check_balance(amount)
      @balance -= amount
    end
  
    # Get the current balance of the bank account
    def balance
      check_status
      @balance
    end
  
    private
  
    # Check the status of the bank account before performing operations
    def check_status(operation = nil)
      if (operation.nil? && ['uninitialized', 'closed'].include?(@status)) ||
         (operation == 'open' && !['uninitialized', 'closed'].include?(@status)) ||
         (operation == 'close' && @status != 'open')
        raise ArgumentError, "Invalid operation: #{@status}"
      end
    end
  
    # Check if there are sufficient funds for withdrawal
    def check_balance(amount)
      raise ArgumentError, "Insufficient funds" if @balance < amount
    end
  
    # Check if the amount is valid (non-negative)
    def check_amount(amount)
      raise ArgumentError, "Invalid amount" if amount < 0
    end
  
  end
